### ISSUE TITLE
Gosh cmd pipelines don't have expected failure behavior

### ISSUE THREAD

**Issue by tatatodd** [vanadium/issues#1082]
As a reminder, if a user wants to run a cmd pipeline "a | b | c", we tell them to do this:

  A, B, C := sh.Cmd("a"), sh.Cmd("b"), sh.Cmd("c")
  A.AddStdoutWriter(B.StdinPipe())
  B.AddStdoutWriter(C.StdinPipe())
  A.Start()
  B.Start()
  C.Start()
  C.Wait()

Our current advice is that they only need to wait on C.  We've coded A.AddStdoutWriter to close B.StdinPipe when A exits, and ditto for B and C.  So in a typical success case, after C.Wait() returns, all cmds A, B and C have exited with success.

What happens if some of the cmds fail?  If C fails, we'll notice the failure with C.Wait().  But if A or B fails, we'll silently ignore the failure, since we've never called Wait() on A or B.  I don't think that's what typical users will want.

I read through "man bash", which details exactly how bash behaves.  By default bash pipelines behave like our current gosh behavior; the exit status of a pipeline is the exit status of the final cmd.  But I don't think that's what typical users want, so people recommend setting "pipefail" to ensure any failure causes the pipeline to fail: https://sipb.mit.edu/doc/safe-shell/

I think that we should add a gosh.Pipeline concept to handle the pipelining explicitly, rather than asking the user to chain Cmds together.  There are many options for what the API will look like.  Ignoring the API, semantically speaking, the gosh.Pipeline will handle:
1. Connecting the stdout/stderr and stdin of adjacent cmds.
2. Running all cmds in the pipeline, handling failures in any of them, like the bash pipefail option.
3. Ensuring that all cmds in the pipeline have exited when the pipeline has finished.

Note that (3) is debateable.  E.g. what happens if C exits before A and B?  Should A and B be killed?  Bash waits for all cmds to exit for a pipeline to finish, and doesn't explicitly kill any of the cmds.  We can discuss this topic separately.

The good news is that if we add gosh.Pipeline, we can make some cleanups.
1. We can change AddStdoutWriter to take an io.Writer, and not have the close-on-exit feature.
2. We can remove gosh.NopWriteCloser, and users can just call AddStdoutWriter(os.Stdout)
3. If we really want some on-exit behavior, we should just add Cmd.AddToExit(f func()).  E.g. we could use that to call PrefixLineWriter.Flush.  This is more explicit, and more general, than our current close-on-exit behavior.

Also a related point.  Currently gosh calls Cmd.closers in our goroutine that's calling exec.Cmd.Wait().  If we add Cmd.AddToExit, this would be weird; users shouldn't need to write thread-safe code for their exit functions.  Instead I think we should ensure that Cmd.closers (and exit funcs) should always be called in gosh.Cmd.wait(), so that we always run them synchronously.  I'm not sure whether this is always possible, due to our Shell onTerminationSignal behavior, but would be nice.

FYI I'm going to lunch now, but if you'd like to chat about this we can have a VC in the afternoon, or we can wait until we're back in the office Tuesday.

**Comment by asadovsky** [comment]
First: what an outstanding issue description! :)
Second: yeah, that's a great point about pipefail, and I didn't think of it before. I like the gist of this proposal, and I think hashing it out in person next Tuesday sounds good. (The cleanups that follow adding gosh.Pipeline indeed sound nice, though I can't remember off the top of my head whether they break any other assumptions; we could try changing the behavior and see which tests fail.)

Btw, I'm curious - what made you think of all this? Are you writing gosh-based command pipelines?

Also, as a side point: as we think about the gosh.Pipeline API, it might be interesting to also consider what a "sequence" or more general "command flow" API might look like, because I could see these APIs being similar. One can think of a pipeline as a sequence where the std streams are connected.

**Comment by tatatodd** [comment]
I don't actually have a gosh-based command pipeline use-case.  So originally I didn't want to spend too much time thinking about pipelines.  But once I realized that this would enable clean-up of AddStdoutWriter and NopWriteCloser, I figured we should work this out.

I started thinking about this stuff after our discussion about Wait() wrt the Terminate() call yesterday.  In particular, I realized that we ignore the exit status of any cmd we don't call Wait() on, and then remembered that we tell people to only Wait() on the last cmd in a pipeline.  I had a 1.5 hour drive through traffic into SF for dinner yesterday, which gave me plenty of thinking time.  :)

Good idea about considering the "sequence" use-case as well.  Let's discuss on Tuesday.
